stumpy/motifs.py

```diff
@@ -0,0 +1,412 @@
+
+    M_T : ndarray, default None
+        Sliding mean of time series, `T`
+
+    Σ_T : ndarray, default None
+        Sliding standard deviation of time series, `T`
+
+    excl_zone : int, default None
+        Size of the exclusion zone.
+        If `None`, defaults to `m/4`, where `m` is the length of `Q`.
+
+    max_distance : float or function, default None
+        Maximum distance between `Q` and a subsequence `S` for `S` to be considered a
+        match.
+        If a function, then it has to be a function of one argument `D`, which will be
+        the distance profile of `Q` with `T` (a 1D numpy array of size `n-m+1`).
+        If None, defaults to `max(np.mean(D) - 2 * np.std(D), np.min(D))`, i.e. at
+        least the closest match will be returned.
+
+    max_matches : int, default None
+        The maximum amount of similar occurrences to be returned. The resulting
+        occurrences are sorted by distance, so a value of `10` means that the
+        indices of the most similar `10` subsequences is returned. If `None`, then all
+        occurrences are returned.
+
+    normalize : bool, default True
+        When set to `True`, this z-normalizes subsequences prior to computing distances.
+        Otherwise, this function gets re-routed to its complementary non-normalized
+        equivalent set in the `@core.non_normalized` function decorator.
+
+    Returns
+    -------
+    out : ndarray
+        The first column consists of indices of subsequences of `T` whose distances
+        to `Q` are smaller than `atol + rtol * profile_value`, sorted by distance
+        (lowest to highest). The second column consist of the corresponding distances.
+    """
+    if len(Q.shape) == 1:
+        Q = Q[np.newaxis, :]
+    if len(T.shape) == 1:
+        T = T[np.newaxis, :]
+
+    d, n = T.shape
+    m = Q.shape[1]
+
+    if excl_zone is None:  # pragma: no cover
+        excl_zone = int(np.ceil(m / 4))
+    if max_matches is None:  # pragma: no cover
+        max_matches = np.inf
+
+    if np.any(np.isnan(Q)) or np.any(np.isinf(Q)):  # pragma: no cover
+        raise ValueError("Q contains illegal values (NaN or inf)")
+
+    if excl_zone is None:  # pragma: no cover
+        excl_zone = int(np.ceil(m / 4))
+
+    if max_distance is None:  # pragma: no cover
+
+        def max_distance(D):
+            return max(np.mean(D) - 2 * np.std(D), np.min(D))
+
+    if M_T is None or Σ_T is None:  # pragma: no cover
+        T, M_T, Σ_T = core.preprocess(T, m)
+
+    if normalize:
+        D = [core.mass(Q[i], T[i], M_T[i], Σ_T[i]) for i in range(d)]
+    else:
+        D = [core.mass_absolute(Q[i], T[i]) for i in range(d)]
+
+    D = np.sum(D, axis=0) / d
+    if not isinstance(max_distance, float):
+        max_distance = max_distance(D)
+
+    matches = []
+
+    candidate_idx = np.argmin(D)
+    while D[candidate_idx] <= max_distance and len(matches) < max_matches:
+        matches.append([candidate_idx, D[candidate_idx]])
```

Comment:
To be consistent with matrix profiles, I feel like we should make the first column the distance and the second column the indices.